src/piton/featurizers/core.py

```diff
@@ -0,0 +1,420 @@
+        )
+        tasks = [
+            (database_path, patient_ids, labeled_patients, self.featurizers)
+            for patient_ids in patient_ids_per_thread
+        ]
+
+        # Preprocess in parallel
+        ctx = multiprocessing.get_context("forkserver")
+        with ctx.Pool(num_threads) as pool:
+            trained_featurizers: List[Featurizer] = [
+                y
+                for x in pool.imap(_run_preprocess_featurizers, tasks)
+                for y in x
+            ]
+        grouped_featurizers: collections.defaultdict = collections.defaultdict(
+            list
+        )
+        for featurizer in trained_featurizers:
+            featurizer_name = featurizer.__class__.__name__
+            grouped_featurizers[featurizer_name].append(featurizer)
+
+        # Aggregating featurizers
+        for idx, featurizer in enumerate(self.featurizers):
+            featurizer_name = featurizer.__class__.__name__
+            self.featurizers[idx] = featurizer.aggregate_featurizers(
+                grouped_featurizers[featurizer_name]
+            )
+
+        for featurizer in self.featurizers:
+            featurizer.finalize_preprocessing()
+
+    def featurize(
+        self,
+        database_path: str,
+        labeled_patients: LabeledPatients,
+        num_threads: int = 1,
+    ) -> Tuple[Any, Any, Any, Any]:
+        """
+        Apply a list of Featurizers (in sequence) to obtain a feature matrix for each Label for each patient.
+
+        Args:
+            database_path (str): Path to `PatientDatabase` on disk
+
+        Returns:
+            This returns a tuple (data_matrix, labels, patient_ids, labeling_time).
+                data_matrix is a sparse matrix of all the features of all the featurizers.
+                label_pids is a list of the patient ids for each row.
+                label_values is a list of boolean values representing the labels for each row in the matrix.
+                labeling_time is a list of labeling/prediction time for each row.
+        """
+
+        patient_ids: List[int] = labeled_patients.get_all_patient_ids()
+        patient_ids_per_thread: List[NDArray[np.int64]] = np.array_split(
+            patient_ids, num_threads
+        )
+        tasks = [
+            (database_path, patient_ids, labeled_patients, self.featurizers)
+            for patient_ids in patient_ids_per_thread
+        ]
+
+        # Run featurizers in parallel
+        ctx = multiprocessing.get_context("forkserver")
+        with ctx.Pool(num_threads) as pool:
+            results: List[Tuple[Any, Any, Any, Any]] = list(
+                pool.imap(_run_featurizer, tasks)
+            )
+
+        # Join results
+        data_matrix = scipy.sparse.vstack([x[0] for x in results])
+        label_pids: NDArray[Literal["n_labels, 1"], np.int64] = np.concatenate(
+            [x[1] for x in results]
+        )
+        label_values: NDArray[Literal["n_labels, 1"], Any] = np.concatenate(
```

Comment:
Not sure this code would work for SurvivalValues? Does np.concatenate work with arbitrary Python objects?